axum/CHANGELOG.md

```diff
@@ -7,11 +7,417 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  let app = Router::new().fallback(fallback);
+
+  async fn fallback() {}
+  ```
+
+- **changed:** `Router::nest` now only accepts `Router`s, the general-purpose
+  `Service` nesting method has been renamed to `nest_service` ([#1368])
+- **breaking:** Allow `Error: Into<Infallible>` for `Route::{layer, route_layer}` ([#924])
+- **breaking:** `MethodRouter` now panics on overlapping routes ([#1102])
+- **breaking:** `Router::route` now only accepts `MethodRouter`s created with
+  `get`, `post`, etc. Use the new `Router::route_service` for routing to
+  any `Service`s ([#1155])
+- **breaking:** Adding a `.route_layer` onto a `Router` or `MethodRouter`
+  without any routes will now result in a panic. Previously, this just did
+  nothing. [#1327]
 - **breaking:** `RouterService` has been removed since `Router` now implements
   `Service` when the state is `()`. Use `Router::with_state` to provide the
-  state and get a `Router<()>` ([#1552])
+  state and get a `Router<()>`. Note that `RouterService` only existed in the
+  pre-releases, not 0.5 ([#1552])
+
+## Extractors
+
+- **added:** Added new type safe `State` extractor. This can be used with
+  `Router::with_state` and gives compile errors for missing states, whereas
+  `Extension` would result in runtime errors ([#1155])
+
+  We recommend migrating from `Extension` to `State` since that is more type
+  safe and faster. That is done by using `Router::with_state` and `State`.
+
+  This setup in 0.5
+
+  ```rust
+  use axum::{routing::get, Extension, Router};
+
+  let app = Router::new()
+      .route("/", get(handler))
+      .layer(Extension(AppState {}));
+  
+  async fn handler(Extension(app_state): Extension<AppState>) {}
+  
+  #[derive(Clone)]
+  struct AppState {}
+  ```
+
+  Becomes this in 0.6 using `State`:
+
+  ```rust
+  use axum::{routing::get, extract::State, Router};
+
+  let app = Router::new()
+      .route("/", get(handler))
+      .with_state(AppState {});
+  
+  async fn handler(State(app_state): State<AppState>) {}
+  
+  #[derive(Clone)]
+  struct AppState {}
+  ```
+
+  If you have multiple extensions you can use fields on `AppState` and implement
+  `FromRef`:
+
+  ```rust
+  use axum::{extract::{State, FromRef}, routing::get, Router};
+
+  let state = AppState {
+      client: HttpClient {},
+      database: Database {},
+  };
+
+  let app = Router::new().route("/", get(handler)).with_state(state);
+  
+  async fn handler(
+      State(client): State<HttpClient>,
+      State(database): State<Database>,
+  ) {}
+  
+  // the derive requires enabling the "macros" feature
+  #[derive(Clone, FromRef)]
+  struct AppState {
+      client: HttpClient,
+      database: Database,
+  }
+  
+  #[derive(Clone)]
+  struct HttpClient {}
+  
+  #[derive(Clone)]
+  struct Database {}
+  ```
+
+- **breaking:** It is now only possible for one extractor per handler to consume
+  the request body. In 0.5 doing so would result in runtime errors but in 0.6 it
+  is a compile error ([#1272])
```

Comment:
I remember one person was relying on content-type checks to make

```rust
fn handler(
    Json(json_payload): Json<Payload>,
    Multipart(multipart): Multipart,
) -> ... {
    ...
}
```

work. Should we add a note that people using this pattern should implement their own extractor like in the `parse-body-based-on-content-type` example?